java/tests/integrationtests/com/google/idea/blaze/java/run/producers/MultipleJavaClassesTestContextProviderTest.java

```diff
@@ -77,17 +75,17 @@ public final void addSourceFolder() {
 
     // required for IntelliJ to recognize annotations, JUnit version, etc.
     workspace.createPsiFile(
-        new WorkspacePath("org/junit/runner/RunWith.java"),
+        new WorkspacePath("java/org/junit/runner/RunWith.java"),
```

Comment:
I was a bit puzzled that the "java/" prefix needed to be added here and below. After all, these JUnit files only need to be present in the PSI so that IntelliJ can resolve them when they are used in the example classes in the tests. Those JUnit files don't represent actual sources of the example project and don't need to be in the `java` source folder.

After seeing this, I dug a bit deeper and noticed that the `clearSourceFolders()` call below removes also the default source folder, meaning that later test runs won't start with the same configuration as the first executed test. Due to this, IntelliJ can't properly resolve the JUnit files anymore (which were apparently automatically placed in the default source folder).

I'm not convinced that calling `clearSourceFolders()` is the right thing to do. If we keep it, we'll at least need a comment for the JUnit classes explaining why they are explicitly added to the `java` source folder (compared to other tests which don't do this).

An alternative could be to not remove the source folder after tests as that won't remove the content anyway. Simply overwriting the `java` source folder at the beginning of each tests (if that folder still exists) seems to work.

Reply:
I noticed that this test is very flaky and my investigation let to [these two lines of code](https://github.com/JetBrains/intellij-community/blob/2f53252d6cfcbb7f258fd4f9e151f662594b1797/java/execution/impl/src/com/intellij/execution/junit/JUnitUtil.java#L250-L251) where IntelliJ fails to find `org.junit.Test`. My theory is it's not finding the class because it's not in a source root and just added to the project.

I decided to go with this approach because it seems cleaner to me. There is no need to clean only java source root but rather to remove all roots. Anyways the JUnit classes are added before each test case.

On other note all other classes that add `org/junit/runner/*.java` files don't add source roots like `MultipleJavaClassesTestContextProviderTest`. I guess I can add a comment here on why this case is a special case.

Reply:
Added in 8769d07c2c3ac9662e413ad99f8954bdcd6793d6.